compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp

```diff
@@ -692,55 +487,357 @@ LogicalResult controlCodeToAie(IRRewriter &rewriter,
+        sourceObjFifo.getCopyLikeProducers();
+    SmallVector<CopyOpInterface> objFifoConsumers =
+        sourceObjFifo.getCopyLikeConsumers();
+    // Default acquire/release value is 1. Will be adjusted depending on number
+    // of producers/consumers.
+    int acqNum{1};
+    if (objFifoConsumers.size() < objFifoProducers.size()) {
+      assert(objFifoProducers.size() % objFifoConsumers.size() == 0);
+      acqNum = objFifoProducers.size() / objFifoConsumers.size();
+    }
+    for (AMDAIE::ChannelOp channel : producerChannels) {
+      Operation *memOp = tileToMemOpMap.at(channel.getTile());
+      AMDAIE::TileOp tileOp = channel.getTileOp();
+      SmallVector<AIE::BufferOp> buffers = llvm::map_to_vector(
+          sourceObjFifo.getBuffersOnTile(tileOp),
+          [&](AMDAIE::BufferOp bufferOp) {
+            return cast<AIE::BufferOp>(mapper.lookup(bufferOp.getOperation()));
+          });
+      SmallVector<AIE::LockOp> producerLocks = llvm::map_to_vector(
+          sourceObjFifo.getProducerLocksOnTile(tileOp),
+          [&](AMDAIE::LockOp lockOp) {
+            return cast<AIE::LockOp>(mapper.lookup(lockOp.getOperation()));
+          });
+      SmallVector<AIE::LockOp> consumerLocks = llvm::map_to_vector(
+          sourceObjFifo.getConsumerLocksOnTile(tileOp),
+          [&](AMDAIE::LockOp lockOp) {
+            return cast<AIE::LockOp>(mapper.lookup(lockOp.getOperation()));
+          });
+      if (producerLocks.size() != 1) {
+        return sourceObjFifo.emitOpError()
+               << "expected a single producer lock for tile: "
+               << channel.getTile() << ", channel: " << channel.getResult();
+      }
+      if (consumerLocks.size() != 1) {
+        return sourceObjFifo.emitOpError()
+               << "expected a single consumer lock for tile: "
+               << channel.getTile() << ", channel: " << channel.getResult();
+      }
+      std::pair<AIE::LockOp, AIE::LockOp> lockPair =
+          std::make_pair(consumerLocks[0], producerLocks[0]);
+      rewriter.moveOpBefore(memOp, deviceBlock, deviceBlock->end());
+      createDMA(memOp, AIE::DMAChannelDir::MM2S, channel.getValue(), dims,
+                acqNum, acqNum, maybeSize.value(), maybeOffset.value(), buffers,
+                lockPair);
+    }
+  }
+
+  SmallVector<Operation *> targetMemOps;
+  Value target = connectionOp.getTarget();
+  auto targetObjFifoLikeOp =
+      dyn_cast_if_present<AMDAIE::LogicalObjFifoOpInterface>(
+          target.getDefiningOp());
+  if (!targetObjFifoLikeOp) {
+    return connectionOp.emitOpError()
+           << "expected target to be an logical objFifo-like op";
+  }
+  if (targetObjFifoLikeOp.getMemorySpaceAsUInt() == 0) {
+    for (AMDAIE::ChannelOp channel : consumerChannels) {
+      auto tileOp = channel.getTileOp();
+      auto shimDmaAllocOp = rewriter.create<AIE::ShimDMAAllocationOp>(
+          rewriter.getUnknownLoc(), "shim_" + std::to_string(connectionIndex++),
```

Comment:
There should be a cleaner way to do this for both source and target ?
How about creating a struct which we fill in (eg: `mixedSourceSizes()` vs `mixedTargetSizes()`, `MM2S` vs `S2MM`, etc) before invoking the function ?

Reply:
Oh okay, makes sense. No worries.